### PR TITLE
Add support for KFServing and CI/test improvements

### PR DESCRIPTION
**Pull request by flaviodsr**
This PR includes the following changes:

- Add support for serving trained models using kfserving
- Add make targets for installing kfserving and its requirements
- Fixes some issues causing instability on the tests/CI and add more verbosity to it

This PR also reverts commit: https://github.com/fuseml/fuseml/commit/3937323ca9d0a2c1e718e80e85e84fc99e953be3 as it seems to not be needed.

### CODE REVIEW

File: scripts/kfserving-install.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+if kubectl get svc kfserving-controller-manager-service -n kfserving-system >/dev/null 2>&1 ; then
+  echo "kfserving already installed"
+  exit
+fi
+
+export KFSERVING_VERSION=v0.5.1
+
+curl https://raw.githubusercontent.com/kubeflow/kfserving/${KFSERVING_VERSION}/install/${KFSERVING_VERSION}/kfserving.yaml | \
+  sed 's/cluster-local/knative-local/' | kubectl apply -f -
```
Comment (jsuchome) [inline scripts/kfserving-install.sh:11]: what is this for?

Comment (flaviodsr) [reply scripts/kfserving-install.sh:11]: kfserving expects the knative gateways to be named as `cluster-local`, however with the newer versions of knative its name changed to `knative-local`

File: embedded-files/tekton/triggers.yaml [hunk 2]
```diff
@@ -214,11 +234,28 @@ spec:
         secretKeyRef:
           name: mlflow-minio
           key: secretkey
-  - name: run-id-to-results
+  - name: model-uri-to-results
```
Comment (jsuchome) [inline embedded-files/tekton/triggers.yaml:237]: does it make sense to have this in separate step? I think model-uri result could be already produced in the run step, is there a reason to start a new container?

File: embedded-files/tekton/triggers.yaml [hunk 3]
```diff
@@ -229,12 +266,74 @@ metadata:
...
   workspaces:
   - name: source
   steps:
   - name: run
     image: lachlanevenson/k8s-kubectl
     workingDir: "/workspace/source/app"
     script: |
-      sed "s/#RUN_ID#/$(params.RUN-ID)/g; s/#IMAGE_SHA#/$(params.IMAGE-SHA)/g" .fuseml/serve.yaml | kubectl apply -f -
+      #!/bin/sh
+      sed "s#__MODEL_URI__#$(params.MODEL-URI)#g; s#__IMAGE_SHA__#$(params.IMAGE-SHA)#g; s#__AWS_ACCESS_KEY_ID__#${AWS_ACCESS_KEY_ID}#g; s#__AWS_SECRET_ACCESS_KEY__#${AWS_SECRET_ACCESS_KEY}#g" .fuseml/serve.yaml | kubectl apply -f -
```
Comment (jsuchome) [inline embedded-files/tekton/triggers.yaml:278]: just wondering - is it safe to use # delimiter for IMAGE_SHA and S3 access keys?

File: embedded-files/tekton/triggers.yaml (hunk 2 above)
Comment (flaviodsr) [reply embedded-files/tekton/triggers.yaml:237]: I have seen this pattern used in tekton and I decided to follow it. I think it makes sense as each step should have its own goal and in case of failure its easier to debug (e.g. the error was getting the URI or training the model?)

File: embedded-files/tekton/triggers.yaml (hunk 3 above)
Comment (flaviodsr) [reply embedded-files/tekton/triggers.yaml:278]: I think its ok. SHA is based on hexadecimal and the s3 keys are alphanumeric.

File: cmd/root.go [hunk 4]
```diff
@@ -51,6 +51,8 @@ func Execute() {
 	viper.BindPFlag("verbosity", pf.Lookup("verbosity"))
 	argToEnv["verbosity"] = "VERBOSITY"
 
+	client.CmdPush.Flags().StringVarP(&client.FlagServe, "serve", "s", "", "inference service to serve the model (kfserving, knative, deployment)")
```
Comment (jsuchome) [inline cmd/root.go:54]: I think the `deployment` option is confusing. Actually `knative` too. They both serve through mlflow so the serve option should really be `mlflow`. If you still want to expose the option to use knative/non-knative way, it should be another argument.

**Comment by jsuchome** [comment]
Please change the PR label, this big PR is obviously about much more.

File: cmd/root.go (hunk 4 above)
Comment (flaviodsr) [reply cmd/root.go:54]: I partially agree, I do not think that mlflow is at the same level as kfserving/seldon. For example, if I choose kfserving the serving is done through mlserver and if I choose seldon the serving can also be done with mlflow or other tool.

In that way I think deployment/knative as the kind of serving is more close to be in the same level as kfserving/seldon.

Comment (jsuchome) [reply cmd/root.go:54]: Hm, I think it will be confusing for users. But we can leave this discussion for later.

**Review by jsuchome — APPROVED** [review]
I like it. 

I would still change the values for "serve" argument, but that's probably for some longer discussion, so no need to wait.

File: cmd/root.go (hunk 4 above)
Comment (stefannica) [reply cmd/root.go:54]: My 2c: there's an option to choose the tool used for serving: `knative`, `kfserving`, `seldon` or `plain kubernetes` (deployment), and there's a sub-option specific to each tool used to choose the serving prediction mechanism. The sub-options could be (for now): `mlflow` (use mlflow instead of a builtin prediction server), `detect` (detect the predictor type based on the library used to train/save the model) or enforce one of the supported libraries (sklearn, pytorch, tensorflow, etc).

Comment (stefannica) [reply cmd/root.go:54]: My suggestions is based on the assumption that we will support tool specific recipes that users can choose from to implement a particular MLOps segment. In this case, these are all serving segments: `knative`, `kfserving`, `seldon` etc. Then there are some generic parameters that all serving segments must support (e.g. the model URL, the prediction engine). And some particular parameters that are specific to each tool.

**Review by stefannica — APPROVED** [review]
